src/runtime_src/core/common/bo_cache.h

```diff
@@ -0,0 +1,90 @@
+/**
+ * Copyright (C) 2019 Xilinx, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"). You may
+ * not use this file except in compliance with the License. A copy of the
+ * License is located at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+#ifndef core_common_bo_cache_h_
+#define core_common_bo_cache_h_
+
+#include <stack>
+#include <utility>
+
+#include "xrt.h"
+#include "ert.h"
+
+namespace xrt_core {
+    // Helper typedef for std::pair. Note the elements are const so that the
+    // pair is immutable. The clients should not change the contents of cmd_bo.
+    typedef std::pair<const unsigned, struct ert_start_copybo_cmd * const> cmd_bo;
+
+    // Create a cache of CMD BO objects -- for now only used for M2M -- to reduce
+    // the overhead of BO cycle management.
+    class bo_cache {
+        xclDeviceHandle device_handle;
+        // Maximum number of BOs that can be cached in the pool
+        unsigned cmd_bo_cache_max_size;
+        std::stack<cmd_bo> cmd_bo_cache;
```

Comment:
std::vector will do just fine, no need for stack.   The default container for stack is deque, which is probably not what you want here.